### ISSUE TITLE
Improper handling of symlinked directories

### ISSUE THREAD

**Issue by Cireo** [PyCQA/pylint#7456]
### Bug description

The behavior is a bit involved, even after significant minimization from the actual problem in my environment, which involves a few orders of magnitude more files.

Symlinked source directories using relative imports leak state (similar to pythonpath) of the symlink destination when other symlinked directories at a different level are also checked in the same pylint command.

Please look at the reproduction script for counter examples.

### Configuration

```ini
$ cat reproduce.sh 
```
gives
```
#!/bin/bash
# All names are arbitrary except base-contents/same must match base-contents/.../same
#
# $ ./reproduce.sh
# pylint 2.13.9
# astroid 2.11.7
# Python 3.6.9 (default, Jan 26 2021, 15:33:00) 
# [GCC 8.4.0]
# Using temporary directory /tmp/tmp.XrKoClgvr7
# No difference between symlink and copy contents
# Run from base-symlink
# ************* Module same
# same/__init__.py:1:0: E0401: Unable to import 'same.baz' (import-error)
# Run from base-copy
# Add invisible base-contents/bar/__init__.py
# Run from base-symlink again
# Remove invisible base-contents/same
# Run from base-symlink again
# Only check same (not wef)
# Run from base-symlink again
#
# $ tree /tmp/tmp.XrKoClgvr7
# .
# ├── base-contents
# │   ├── bar
# │   │   ├── __init__.py (SOMETIMES)
# │   │   └── same
# │   │       ├── baz.py
# │   │       └── __init__.py
# │   ├── same (SOMETIMES)
# │   │   └── __init__.py
# │   └── wef
# │       └── __init__.py
# ├── base-copy
# │   ├── same
# │   │   ├── baz.py
# │   │   └── __init__.py
# │   └── wef
# │       └── __init__.py
# └── base-symlink
#     ├── same -> ../base-contents/bar/same
#     └── wef -> ../base-contents/wef

python -m pylint --version

TMPDIR=$(mktemp -d)
echo "Using temporary directory $TMPDIR"
cd "$TMPDIR"

mkdir base-contents
mkdir -p base-contents/bar/same
echo 'val = 1' > base-contents/bar/same/baz.py
echo 'from .baz import val' > base-contents/bar/same/__init__.py
mkdir base-contents/same
touch base-contents/same/__init__.py
mkdir base-contents/wef
touch base-contents/wef/__init__.py

mkdir base-symlink
ln -s ../base-contents/bar/same base-symlink/same
ln -s ../base-contents/wef base-symlink/wef

mkdir base-copy
cp -R base-contents/bar/same base-copy/same
cp -R base-contents/wef base-copy/wef

echo "No difference between symlink and copy contents"
diff -r base-copy/ base-symlink/

# Fails
# ************* Module same
# same/__init__.py:1:0: E0401: Unable to import 'same.baz' (import-error)
echo "Run from base-symlink"
(cd base-symlink && python -m pylint -E wef/__init__.py same/__init__.py)

# Passes
echo "Run from base-copy"
(cd base-copy && python -m pylint -E wef/__init__.py same/__init__.py)

# Passes
echo "Add invisible base-contents/bar/__init__.py"
touch base-contents/bar/__init__.py
echo "Run from base-symlink again"
(cd base-symlink && python -m pylint -E wef/__init__.py same/__init__.py)
rm base-contents/bar/__init__.py

# Passes
echo "Remove invisible base-contents/same"
mv base-contents/{same,not_same}
echo "Run from base-symlink again"
(cd base-symlink && python -m pylint -E wef/__init__.py same/__init__.py)
mv base-contents/{not_same,same}

# Passes
echo "Only check same (not wef)"
echo "Run from base-symlink again"
(cd base-symlink && python -m pylint -E same/__init__.py)
```
```


### Command used

```shell
python -m pylint -E wef/__init__.py same/__init__.py
```
```


### Pylint output

```shell
************* Module same
same/__init__.py:1:0: E0401: Unable to import 'same.baz' (import-error)
```
```


### Expected behavior

```
```

### Pylint version

```shell
# pylint 2.13.9
# astroid 2.11.7
# Python 3.6.9 (default, Jan 26 2021, 15:33:00) 
```
```


### OS / Environment

Ubuntu 18.04

### Additional dependencies

_No response_

**Comment by jacobtylerwalls** [comment]
Can you reproduce against either `pylint` 2.15.2 or `main`?

**Comment by Cireo** [comment]
For python 3.6.9, 2.13.9 (what I ran above) is the latest available version
```
# pylint 2.13.9
# astroid 2.11.7
# Python 3.6.9 (default, Jan 26 2021, 15:33:00) 
```
when I switch to python 3.9.0, 2.15.2 is available, and I cannot reproduce
```
$ ./reproduce.sh 
pylint 2.15.2
astroid 2.12.9
Python 3.9.0 (default, Oct 18 2020, 15:19:13) 
[Clang 12.0.0 (clang-1200.0.32.2)]
Using temporary directory /var/folders/ht/gy62n83n2f9009b2fynly9tm0000gn/T/tmp.SUZ9nksr
No difference between symlink and copy contents
Run from base-symlink
Run from base-copy
Add invisible base-contents/bar/__init__.py
Run from base-symlink again
Remove invisible base-contents/same
Run from base-symlink again
Only check same (not wef)
Run from base-symlink again
```

**Comment by DanielNoord** [comment]
We long support Python `3.6` and don't provide bug fix updates to the `2.13.x` versions. Sorry about that!

I'd advise updating to a new Python interpreter as `3.6` is past EOL and other tools have likely also dropped support.
